File: Data Structures & Algorithms/longest-substring-without-duplicates/submission-3.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        unique = set()
        mc = 0
        l, r = 0, 0

        while r < len(s):
            while s[r] in unique:
                    unique.remove(s[l])
                    l += 1
            unique.add(s[r])
            r += 1
            mc = max(mc, r - l)
        return mc

File: Data Structures & Algorithms/trapping-rain-water/submission-3.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int res = 0;
        int left = 0, right = height.size() - 1;
        int left_max = height[0], right_max = height[right];
        while (left < right) {
            if (height[left] < height[right]) {
                left++;
                left_max = max(left_max, height[left]);
                res += left_max - height[left];
            } else {
                right--;
                right_max = max(right_max, height[right]);
                res += right_max - height[right];
            }
        }
        return res;
    }
};
